### DISCUSSION TITLE
Literature

### DISCUSSION THREAD

**Discussion by jejjohnson** [IPL-UV/rbig_jax#27] [Resources]
# Convolutions (Orthogonal)

**[Glow: Generative flow with invertible 1x1 convolutions](https://arxiv.org/abs/1807.03039)** - Kingma & Dhariwal (2018)

> Original paper showcasing a 1x1 Invertible Convolution

* **Problem** 1: It's invertible but it is not orthogonal. So it violates some of the problems with the original RBIG construction and the relationship between the KLD and information.
* **Problem** 2: I don't think it account for the spatial variability. I could be wrong but it may only be channel-wise. They mentioned it in the *convolutional exponential paper*.

---

**[Emerging convolutions for generative normalizing flows](https://arxiv.org/abs/1901.11137)** - Hoogeboom et al (2019)

> A paper that creating an orthogonal convolutional which **does** account for the spatial characteristics. They also improve the original 1x1Conv Method by parameterized it with householder rotations. So it makes it correspond with RBIG convergence specifics.

* **Problem** 1: It is fast for evaluating densities, but it is slow for generating samples. It uses a sort of autoregressive approach.

---

**[The Convolution Exponential and Generalized Sylvester Flows](https://arxiv.org/abs/2006.01910)** - Hoogeboom et al (2020)

> They create the *exponential* convolution transform. It basically uses the exponential of the convolutions. It approximates the transform via a Taylor series. It generalizes most of the householder approaches, i.e. Sylvester Flows.

* **Problem** 1: I saw in the code that they did some spectral normalization which greatly increases the complexity of the code. I don't understand if it is necessary for our application or maybe it's only for Dequantization.

# MultiScale Architecture

NICE: Non-linear Independent Components Estimation - Dinh et al (2015) - [Paper](https://arxiv.org/abs/1410.8516) | [Code](https://github.com/DakshIdnani/pytorch-nice)

---

Density estimation using Real NVP - Dinh et al (2016) - [Paper](https://arxiv.org/abs/1605.08803) | Code ([JAX](https://nbviewer.jupyter.org/github/glouppe/flowing-with-jax/blob/master/jax-realnvp.ipynb) | [PyTorch](https://nbviewer.jupyter.org/github/senya-ashukha/real-nvp-pytorch/blob/master/real-nvp-pytorch.ipynb)) | [Video](https://www.youtube.com/watch?v=P4Ta-TZPVi0)

# Conditional Density Estimation

Learning Likelihoods with Conditional Normalizing Flows - Winkler et al (2019) - [Paper](https://arxiv.org/abs/1912.00042) )

<details>
<summary> Abstract </summary>
Normalizing Flows (NFs) are able to model complicated distributions p(y) with strong inter-dimensional correlations and high multimodality by transforming a simple base density p(z) through an invertible neural network under the change of variables formula. Such behavior is desirable in multivariate structured prediction tasks, where handcrafted per-pixel loss-based methods inadequately capture strong correlations between output dimensions. We present a study of conditional normalizing flows (CNFs), a class of NFs where the base density to output space mapping is conditioned on an input x, to model conditional densities p(y|x). CNFs are efficient in sampling and inference, they can be trained with a likelihood-based objective, and CNFs, being generative flows, do not suffer from mode collapse or training instabilities. We provide an effective method to train continuous CNFs for binary problems and in particular, we apply these CNFs to super-resolution and vessel segmentation tasks demonstrating competitive performance on standard benchmark datasets in terms of likelihood and conventional metrics.
</details>

---

Semi-Supervised Learning with Normalizing Flows Izmailov et al. (2019) - [Paper](https://arxiv.org/abs/1912.13025) | [Code](https://github.com/izmailovpavel/flowgmm)

---

Noise Regularization for Conditional Density Estimation - Rothfuss et al (2020) - [Paper](https://arxiv.org/abs/1907.08982) | [Code](https://github.com/freelunchtheorem/Conditional_Density_Estimation)

---

Conditional Density Estimation with Bayesian Normalising Flows - Trippe et al (2018) - [Paper](https://arxiv.org/abs/1802.04908) | [Code](https://github.com/blt2114/CDE_with_BNF)

# Wavelets

Wavelet Flow: Fast Training of High Resolution Normalizing Flows - Yu et al (2020) - [Paper](https://arxiv.org/abs/2010.13821) | [Code](https://github.com/YorkUCVIL/Wavelet-Flow/)
